crate2nix/src/resolve.rs

```diff
@@ -616,9 +653,45 @@ impl CratesIoSource {
             version = self.version
         )
     }
+
+    /// Construct a URL for fetching a crate artifact based on the URL of the
+    /// registry index.
+    ///
+    /// For crates.io there is a known download URL we use:
+    ///   https://static.crates.io/crates/{name}/{name}-{version}.crate
+    ///
+    /// For custom registries we assume the index url has the form
```

Comment:
Ideal implementatino would look this up in the registry itself, It's `config.json` contains the pattern from which the download URL is constructed. See https://doc.rust-lang.org/cargo/reference/registries.html#index-format

That would allow to avoid the specific handling for the two registries. If we continue doing this, then IMHO the implementation should replace the mentions of "registry" with "cloudsmith" for which this assumption holds.

Reply:
I couldn't find a config.json on Cloudsmith, so I just updated the code to be explicit in Cloudsmith support, hopefully this will address the concerns.

However I am in favor of keeping the Registry and RegistrySource structs as is, since it unifies the interface for dealing with registries, I don't see the benefit of creating more structs to represent unique registry implementations. (There is nothing idiosyncratic about them)

Reply:
Its definitely there, I cloned the index repo today and it was in the root of the repository.

Reply:
Ah yes you're right it's there. I was looking for a way to access it via raw http.. otherwise I imagine pulling this file down will need a git client bundled with this library. 🤔

Reply:
So I think the current implementation is sufficient for adding cloudsmith support. Will leave generic registry support (by parsing config.json) for a future enhancement.